### PR TITLE
fix: message composer not changing state correctly when hiding keyboard [AR-2720] [AR-2723] [AR-1477]

### PR DESCRIPTION
**Pull request by trOnk12**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2720" title="AR-2720" target="_blank">BugAR-2720</a>  [Message composer] Activating the inputfield works not as expected
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- the message composer was unncessary complicated, that is why I decided to remove nested ConstraintLayouts mixed with Column to make it easier to follow what is happening.

AttachmentOptions and MessageComposerInput are two elements inside the View, this is done to make the UI not flicker when closing and opening keyboard or closing and opening attachment options. 

MessageComposerInput is now simple view following the measure policy of Column 

- Also correctly move the MessageComposer to enabled state when the user decides to close the keyboard

### Attachments (Optional)
Animation

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/1/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt [hunk 1]
```diff
@@ -111,71 +115,64 @@ fun MessageComposer(
...
 
+        LaunchedEffect(keyboardHeight) {
+            messageComposerState.keyboardHeight = keyboardHeight
+        }
+
+        LaunchedEffect(isKeyboardVisible) {
+            if (!isKeyboardVisible && !messageComposerState.attachmentOptionsDisplayed) {
+                messageComposerState.toEnabled()
+                messageComposerState.focusManager.clearFocus()
+                messageComposerState.cancelReply()
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: do we really want to cancel reply when user just hides a keyboard?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 2]
```diff
@@ -373,26 +374,28 @@ private fun ConversationScreen(
...
-                            onReactionClicked = onReactionClick,
-                            onOpenProfile = onOpenProfile,
-                            onUpdateConversationReadDate = onUpdateConversationReadDate,
-                            onMessageComposerError = onSnackbarMessage,
-                            onShowContextMenu = conversationScreenState::showEditContextMenu,
-                            onSnackbarMessageShown = onSnackbarMessageShown,
-                            snackbarMessage = conversationViewState.snackbarMessage ?: conversationMessagesViewState.snackbarMessage
+                            isKeyboardVisible = isKeyboardVisible,
+                        membersToMention = membersToMention,
+                        isFileSharingEnabled = conversationViewState.isFileSharingEnabled,
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:356]: fix the code-formatting plz :)

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt [hunk 3]
```diff
@@ -186,172 +183,254 @@ private fun MessageComposer(
...
+                )
+            }
+        }
+
+    }
+
+    BackHandler(messageComposerState.attachmentOptionsDisplayed) {
+        messageComposerState.hideAttachmentOptions()
+        messageComposerState.toEnabled()
+        messageComposerState.cancelReply()
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: the same here: why do we cancel reply on backPress when attachments are visible?

Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: good question 🤔

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt (hunk 1 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: good question 🤔

File: app/src/main/kotlin/com/wire/android/ui/common/bottomsheet/WireModalSheetLayout.kt [hunk 4]
```diff
@@ -38,14 +40,19 @@ fun WireModalSheetLayout(
...
         sheetContent = {
-            Spacer(modifier = Modifier.height(8.dp))
-            Divider(
-                modifier = Modifier
-                    .width(width = dimensions().modalBottomSheetDividerWidth)
-                    .align(alignment = Alignment.CenterHorizontally),
-                thickness = 4.dp
-            )
-            sheetContent()
+            val alpha =
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/common/bottomsheet/WireModalSheetLayout.kt]: maybe a comment why we are doing it ?

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/3/) **failed**.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5285401)(5285401)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5285401)(5285401)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/5/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5316396)(5316396)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5316396)(5316396)-dev-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/common/bottomsheet/WireModalSheetLayout.kt (hunk 4 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/bottomsheet/WireModalSheetLayout.kt]: I added a comment with explanation

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/8/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5318581)(5318581)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5318581)(5318581)-dev-compat.apk)

**Comment by saleniuk** [comment]
https://user-images.githubusercontent.com/30429749/209684821-84090fad-9c5b-4497-8ad3-32075601eed4.mp4

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt (hunk 1 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: it's removed, it shouldn't cancel the reply

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt (hunk 3 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/MessageComposer.kt]: it's removed, it shouldn't cancel the reply

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/9/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5324988)(5324988)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5324988)(5324988)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [11](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/11/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5326771)(5326771)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5326771)(5326771)-dev-compat.apk)

**Review by saleniuk — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [12](https://10.10.124.134/job/wire-android-reloaded/job/PR-1231/12/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5327889)(5327889)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5327889)(5327889)-dev-compat.apk)

**Pull request merged**